doc/storage-tier-design.md

```diff
@@ -0,0 +1,241 @@
+will be selected for provisioning.
+
+```yaml
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: premium
+parameters:
+  repl: '1'
+  protocol: 'nvmf'
+  ioTimeout: '60'
+provisioner: io.openebs.csi-mayastor
+volumeBindingMode: WaitForFirstConsumer
+allowedTopologies:
+- matchLabelExpressions:
+  - key: openebs.io/classification/premium
+    values:
+    - "true"
+```
+
+### CSI Node Topology
+
+CSI drivers normally register the labels that they use for topology. They do
+this by returning the label key and any currently supported values in the CSI
+`NodeGetInfo` response. For example, they might return:
+
+```protobuf
+message NodeGetInfoResponse {
+  node_id = "csi-node://nodeA";
+  max_volumes_per_node = 16;
+  accessible_topology = {"kubernetes.io/hostname": "nodeA", openebs.io/classification/premium: "true"}
+}
+```
+
+The `NodeGetInfo` request is initiated by kubelet when the
+`csi-driver-registrar` container starts and registers itself as a kubelet
+plugin. It is only called once. It calls the `NodeGetInfo` endpoint on the
+`csi-node` container.
+
+Topology KV pairs returned in the `NodeGetInfoResponse` are added as labels on
+the `Node` object:
+
+```yaml
+apiVersion: v1
+kind: Node
+metadata:
+  annotations:
+    csi.volume.kubernetes.io/nodeid: '{"disk.csi.azure.com":"aks-storagepool-25014071-vmss000000","io.openebs.csi-mayastor":"csi-node://aks-storagepool-25014071-vmss000000"}'
+  labels:
+    openebs.io/engine: mayastor
+    openebs.io/classification/premium: "true"
+  name: aks-storagepool-25014071-vmss000000
+spec:
+  ...
+```
+
+The topology keys are registered against the CSI driver on the `CSINode` object:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSINode
+metadata:
+  name: aks-storagepool-25014071-vmss000000
+spec:
+  drivers:
+  - name: io.openebs.csi-mayastor
+    nodeID: csi-node://aks-storagepool-25014071-vmss000000
+    topologyKeys:
+    - kubernetes.io/hostname
+    - openebs.io/classification/premium
+```
+
+There are two issues with using the CSI driver registration process to manage
+tiering topology:
+
+1. Since `NodeGetInfo` is only called once when `csi-driver-registrar` starts,
+   any changes to node topology (e.g. adding a `DiskPool`) will not be reflected
+   on the node labels. A mechanism to trigger plugin re-registration is needed.
+   This could be achieved by adding a livenessProbe on `csi-driver-registrar`
+   that queries an endpoint that fails if an update is required. This would
+   cause the `csi-driver-registrar` to restart.
+2. The `csi-node` process does not currently have knowledge of the `DiskPools`
+   exposed on the node so it would need to request them from the control plane's
+   REST endpoint so it can be returned in the `NodeGetInfo` response. This would
+   require adding communication from the node to the API, which is not ideal.
+
+Instead, a separate controller is proposed. It will run alongside the diskpool
```

Comment:
At the moment we use csi-node topology to control where the diskpools are placed, which as you've pointed out is not quite right (once we remove the locality constraint the csi-node will be able to run on non-"openebs/engine: mayastor" nodes).

What is the actual intent from CSI for the accessible topology?
```
  // Specifies the list of topologies the provisioned volume MUST be
  // accessible from
```
Since our volumes are accessed via nvme-tcp, maybe this is not intended to control the placement of data replicas?
And in that case, should we separate the application topology and data topology completely?
If so, we could, for example, pass topology information through the storage class's generic parameters and this way we wouldn't need to sync between `diskpools` and `CSINode`?

Reply:
This is good point to consider, we can have two different topologies one is used for placement of replicas & other one is K8s `allowedTopologies`(where data can be accessible from).  

- If we have two different topologies for apps & replica placement then it will serve both use cases:
   - Disaggregated storage use case, where set of nodes in a cluster are dedicated for storage, topology will not influence in pod scheduling.
    - Normal cluster where all the nodes are capable of hosting storage & apps no issues.

Reply:
Good points:
- once the locality requirement is removed, the volume will be available on any node
- separate topologies for apps & replica placement

So in effect, we shouldn't re-use CSI placement topology - as long as the replica topology is passed on the `CreateVolume` call, it can still be stored on the Volume's internal topology and placement works as intended.

This should just mean converting the StorageClass's `allowedTopologies` to KV pairs in the SC parameters.  That should be ok since we don't need any complicated logic.

I'll try this out and update the doc.

Reply:
I tried it out, and I agree this is better:

DiskPool:
```yaml
apiVersion: openebs.io/v1alpha1
kind: DiskPool
metadata:
  labels:
    openebs.io/classification: premium
  name: diskpool-jr4nm
spec:
  disks:
  - /dev/sdc
  node: aks-storagepool-25014071-vmss000000
```
StorageClass:
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: premium
parameters:
  repl: '1'
  protocol: 'nvmf'
  ioTimeout: '60'
  classification: premium
provisioner: io.openebs.csi-mayastor
volumeBindingMode: WaitForFirstConsumer
```
A small code change in csi controller's `create_volume` to pass the parameter in the volume's labelledTopology:

```json
{
    "uuid": "7bb6dbc2-c170-46cd-b627-c0a04d613a24",
    "size": 1073741824,
    "labels": null,
    "num_replicas": 1,
    "status": {
        "Created": "Online"
    },
    "target": {
        "node": "aks-storagepool-25014071-vmss000000",
        "nexus": "7c14d3ff-459e-4aea-967a-c3b4dd47ea12",
        "protocol": "nvmf"
    },
    "policy": {
        "self_heal": true
    },
    "topology": {
        "node": {
            "Explicit": {
                "allowed_nodes": [
                    "aks-nodepool1-14021096-vmss000000",
                    "aks-nodepool1-14021096-vmss000001",
                    "aks-nodepool1-14021096-vmss000002",
                    "aks-storagepool-25014071-vmss000000",
                    "aks-storagepool-25014071-vmss000001",
                    "aks-storagepool-25014071-vmss000002"
                ],
                "preferred_nodes": [
                    "aks-storagepool-25014071-vmss000000",
                    "aks-storagepool-25014071-vmss000001",
                    "aks-storagepool-25014071-vmss000002",
                    "aks-nodepool1-14021096-vmss000000",
                    "aks-nodepool1-14021096-vmss000001",
                    "aks-nodepool1-14021096-vmss000002"
                ]
            }
        },
        "pool": {
            "Labelled": {
                "exclusion": {},
                "inclusion": {
                    "openebs.io/classification": "premium",
                    "openebs.io/created-by": "operator-diskpool"
                }
            }
        }
    },
    "last_nexus_id": "7c14d3ff-459e-4aea-967a-c3b4dd47ea12",
    "operation": null
}

Is this more what you were thinking?
```

Reply:
That's great.
Do you see classification as a reserved key word? I think it'd be good to keep it generic so you can use whatever labels you want?

Reply:
I was thinking it being a reserved key word, but it doesn't need to be.  The issue is knowing which SC parameter to copy from the `CreateVolume` request to the volume.

How about having a default: `openebs.io/classification`, but have an optional SC param for `openebs.io/classification_key` that can override it?  I think it's fine to have this configurable on the SC.  It could also be a flag on the CSI controller but that seems much less flexible.

And since naming is hard...  I don't know whether classification is the right term, but I think it would be better to use the label prefix (openebs.io/) everywhere to avoid confusion.  We definitely want to use the prefix for the labels.

Reply:
> This is good point to consider, we can have two different topologies one is used for placement of replicas & other one is K8s `allowedTopologies`(where data can be accessible from).
> 
> * If we have two different topologies for apps & replicas placement then it will serve both use cases:
>   
>   * In disaggregated storage use case (where set of nodes in a cluster are dedicated for storage and other nodes for application) by following `storageclass.parameters.dataPlacementTopology` can provision replicas based on topology and `AllowedTopology` can be used for accessibility of volume (Where app can schedule).
>   * In normal cluster where all the nodes are capable of hosting both storage & apps then one can place replica depends on `storageclass.parameters.dataPlacementTopology` & application can consume volume from any node.

Since we have agreed upon this, we would not be using `allowed topologies` i.e data-accessibility topology for data-placement. The data-placement topology can be labelled topology incorporated from the storage-class `parameters`. This also gives us benefit of using pools on newly added nodes to cluster as we allowing scaling up of volume after creation, which would otherwise be a problem if we used the explicit topology filled from allowed topology at the time of creation.

https://github.com/openebs/mayastor-control-plane/pull/275 ->  PR to seperate, the accessibilty from data-placement.